### PR TITLE
HPCC-22984 Provide layout mismatch warnings in WU summary

### PR DESCRIPTION
**Pull request by richardkchapman**
Hthor changes to generate workunit information regarding translation.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@jakesmith May need something in Thor?

File: ecl/hthor/hthor.cpp [hunk 1]
```diff
@@ -8144,10 +8145,13 @@ void CHThorDiskReadBaseActivity::ready()
...
                 publishedCrc = props.getPropInt("@formatCrc");
         }
+        dFile->getLogicalName(traceName);
     }
-    translators.setown(::getTranslators("hthor-diskread", expectedCrc, expectedDiskMeta, publishedCrc, publishedMeta, projectedCrc, projectedDiskMeta, translationMode));
+    translators.setown(::getTranslators(traceName.str(), expectedCrc, expectedDiskMeta, publishedCrc, publishedMeta, projectedCrc, projectedDiskMeta, translationMode));
     if (translators)
     {
+        VStringBuffer msg("Record layout translation required for %s", traceName.str());
+        agent.addWuExceptionEx(msg.str(), WRN_UseLayoutTranslation, SeverityInformation, MSGAUD_user, "hthor");
```
Comment (jakesmith) [inline ecl/hthor/hthor.cpp]: I think this is going to triggered for regular compound projects too (which isn't wanted)?

**Comment by jakesmith** [comment]
@richardkchapman - 1 comment/question.

I've opened a separate JIRA for changes to Thor to do same: https://track.hpccsystems.com/browse/HPCC-23087

File: ecl/hthor/hthor.cpp (hunk 1 above)
Comment (richardkchapman) [reply ecl/hthor/hthor.cpp]: Updated to only report cases where actual and expected differ

**Comment by richardkchapman** [comment]
@jakesmith Please re-review

**Comment by jakesmith** [comment]
@richardkchapman - looks good.

**Comment by richardkchapman** [comment]
@ghalliday Please review/merge

File: ecl/hthor/hthorkey.cpp [hunk 2]
```diff
@@ -4187,11 +4188,17 @@ class CHThorKeyedJoinActivity  : public CHThorThreadedActivityBase, implements I
         if (actualFormat)
         {
             actualLayouts.append(actualFormat.getLink());  // ensure adequate lifespan
-            Owned<const IDynamicTransform> payloadTranslator =  createRecordTranslator(helper.queryProjectedIndexRecordSize()->queryRecordAccessor(true), actualFormat->queryRecordAccessor(true));
+            Owned<const IDynamicTransform> payloadTranslator = createRecordTranslator(helper.queryProjectedIndexRecordSize()->queryRecordAccessor(true), actualFormat->queryRecordAccessor(true));
+            DBGLOG("Record layout translator created for %s", f->queryLogicalName());
```
Comment (ghalliday) [inline ecl/hthor/hthorkey.cpp]: What is the rule for when logging should appear, and the translator should be described?  I imagine the output might be quite large.  Does it differ for files and keys?

Comment (ghalliday) [reply ecl/hthor/hthorkey.cpp]: Should it only be if expected != actual, rather than projected != expected?

File: ecl/hthor/hthor.cpp [hunk 3]
```diff
@@ -8144,10 +8145,16 @@ void CHThorDiskReadBaseActivity::ready()
             if (publishedMeta)
                 publishedCrc = props.getPropInt("@formatCrc");
         }
+        dFile->getLogicalName(traceName);
```
Comment (ghalliday) [inline ecl/hthor/hthor.cpp]: should traceName default to hthor-diskread if !dFile?

**Review by ghalliday — COMMENTED** [review]
@richardkchapman  generally looks sensible.  a couple of questions.

File: ecl/hthor/hthorkey.cpp (hunk 2 above)
Comment (richardkchapman) [reply ecl/hthor/hthorkey.cpp]: To have reached this code, actual and expected must differ (we don't have the three layouts in indexes yet)

File: ecl/hthor/hthor.cpp (hunk 3 above)
Comment (richardkchapman) [reply ecl/hthor/hthor.cpp]: Can do

**Comment by richardkchapman** [comment]
@ghalliday Please re-review

**Comment by ghalliday** [comment]
@richardkchapman looks good.  Please squash.

**Comment by richardkchapman** [comment]
@ghalliday Squashed

**Pull request merged**